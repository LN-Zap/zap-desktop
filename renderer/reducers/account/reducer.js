import createReducer from '@zap/utils/createReducer'
import { getIntl } from '@zap/i18n'
import { mainLog } from '@zap/utils/log'
import waitForIpcEvent from '@zap/utils/waitForIpc'
import { closeDialog } from 'reducers/modal'
import { showNotification } from 'reducers/notification'
import accountSelectors from './selectors'
import messages from './messages'
import * as constants from './constants'

const {
  INIT_ACCOUNT,
  INIT_ACCOUNT_SUCCESS,
  INIT_ACCOUNT_FAILURE,
  LOGIN,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGIN_CLEAR_ERROR,
  CHANGE_PASSWORD_DIALOG_ID,
  PASSWORD_PROMPT_DIALOG_ID,
  PASSWORD_SET_DIALOG_ID,
  SET_IS_PASSWORD_ENABLED,
  LOGIN_NOT_ALLOWED,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

export const initialState = {
  isAccountLoading: false,
  isAccountLoaded: false,
  initAccountError: null,
  isLoggingIn: false,
  loginError: null,
  isLoggedIn: false,
  isPasswordEnabled: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * initAccount - Fetch the current account info from the database and save into the store.
 * Should be called once when the app first loads.
 *
 * @returns {Function} Thunk
 */
export const initAccount = () => async dispatch => {
  dispatch({ type: INIT_ACCOUNT })
  try {
    // Auto login user if password feature is disabled.
    const isAccountPasswordEnabled = await dispatch(checkAccountPasswordEnabled())
    if (!isAccountPasswordEnabled) {
      dispatch({ type: LOGIN_SUCCESS })
    }
    dispatch({ type: INIT_ACCOUNT_SUCCESS })
  } catch (e) {
    mainLog.warn('checkAccountPasswordEnabled error: %o', e)
    dispatch({ type: LOGIN_FAILURE, error: LOGIN_NOT_ALLOWED })
  }
}

/**
 * setIsPasswordEnabled - Whether password is set for the account.
 *
 * @param {string} value value
 * @returns {object} Action
 */
const setIsPasswordEnabled = value => ({
  type: SET_IS_PASSWORD_ENABLED,
  value,
})

/**
 * setPassword - Updates wallet password.
 *
 * @param {string} password new password
 * @returns {Function} Thunk
 */
const setPassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  dispatch(waitForIpcEvent('setPassword', { value: await sha256digest(password) }))
}

/**
 * checkAccountPasswordEnabled - Checks whether app password is set via checking secure storage.
 * Dispatches setIsPasswordEnabled on completion.
 *
 * @returns {Function} Thunk
 */
const checkAccountPasswordEnabled = () => async dispatch => {
  const { value } = await dispatch(waitForIpcEvent('hasPassword'))
  dispatch(setIsPasswordEnabled(value))
  return value
}

/**
 * changePassword - Changes existing password.
 *
 * @param {object} params password params
 * @param {string} params.newPassword new password
 * @param {string} params.oldPassword old password
 * @returns {Function} Thunk
 */
export const changePassword = ({ newPassword, oldPassword }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(oldPassword))
    await dispatch(setPassword(newPassword))
    dispatch(closeDialog(CHANGE_PASSWORD_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_updated)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * enablePassword - Enables app-wide password protection.
 *
 * @param {string} password to be used further on.
 * @returns {Function} Thunk
 */
export const enablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(setPassword(password))
    dispatch(closeDialog(PASSWORD_SET_DIALOG_ID))
    dispatch(setIsPasswordEnabled(true))
    dispatch(showNotification(intl.formatMessage(messages.account_password_enabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * disablePassword - Disables app-wide password protection.
 *
 * @param {string} password current password.
 * @returns {Function} Thunk
 */
export const disablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(password))
    await dispatch(waitForIpcEvent('deletePassword'))
    dispatch(setIsPasswordEnabled(false))
    dispatch(closeDialog(PASSWORD_PROMPT_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_disabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * requirePassword - Password protect routine. Should be placed before protected code.
 *
 * @param {string} password current password.
 * @returns {Promise} Promise that fulfills after login attempt (either successful or not)
 */
const requirePassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  const { password: hash } = await dispatch(waitForIpcEvent('getPassword'))
  const passwordHash = await sha256digest(password)
  // compare hash received from the main thread to a hash of a password provided
  if (hash === passwordHash) {
    return true
  }
  throw new Error(getIntl().formatMessage(messages.account_invalid_password))
}

/**
 * login - Perform account login.
 *
 * @param {string} password Password
 * @returns {Function} Thunk
 */
export const login = password => async dispatch => {
  try {
    dispatch({ type: LOGIN, password })
    await dispatch(requirePassword(password))
    dispatch({ type: LOGIN_SUCCESS })
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * clearLoginError - Clear login user error.
 *
 * @returns {Function} Thunk
 */
export const clearLoginError = () => (dispatch, getState) => {
  if (accountSelectors.loginError(getState())) {
    dispatch({
      type: 'LOGIN_CLEAR_ERROR',
    })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [INIT_ACCOUNT]: state => {
    state.isAccountLoading = true
  },
  [INIT_ACCOUNT_SUCCESS]: state => {
    state.isAccountLoading = false
    state.isAccountLoaded = true
    state.initAccountError = null
  },
  [INIT_ACCOUNT_FAILURE]: (state, { error }) => {
    state.isAccountLoading = false
    state.initAccountError = error
  },

  [LOGIN]: state => {
    state.isLoggingIn = true
  },
  [LOGIN_SUCCESS]: state => {
    state.isLoggingIn = false
    state.isLoggedIn = true
    state.loginError = null
  },
  [LOGIN_FAILURE]: (state, { error }) => {
    state.isLoggingIn = false
    state.loginError = error
  },
  [LOGIN_CLEAR_ERROR]: state => {
    state.loginError = null
  },
  [SET_IS_PASSWORD_ENABLED]: (state, { value }) => {
    state.isPasswordEnabled = value
  },
}

// ------------------------------------
// Reducer
// ------------------------------------

export default createReducer(initialState, ACTION_HANDLERS)
