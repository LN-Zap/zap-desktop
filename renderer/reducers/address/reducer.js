import get from 'lodash/get'

import { getIntl } from '@zap/i18n'
import createReducer from '@zap/utils/createReducer'
import { openModal, closeModal } from 'reducers/modal'
import { showError } from 'reducers/notification'
import { settingsSelectors } from 'reducers/settings'
import { grpc } from 'workers'

import messages from '../messages'
import * as constants from './constants'

const {
  FETCH_ADDRESSES,
  FETCH_ADDRESSES_SUCCESS,
  FETCH_ADDRESSES_FAILURE,
  NEW_ADDRESS,
  NEW_ADDRESS_SUCCESS,
  NEW_ADDRESS_FAILURE,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isAddressLoading Boolean indicating if addresses are loading.
 * @property {Error|null} fetchAddressError Error from loading addresses.
 * @property {{np2wkh: boolean, p2wkh: boolean}} addressesLoading Booleans indicating address types that are loading.
 * @property {{np2wkh: string|null, p2wkh: string|null}} addresses Current bitcoin addresses.
 */

/** @type {State} */
const initialState = {
  isAddressLoading: false,
  fetchAddressError: null,
  addressesLoading: {
    np2wkh: false,
    p2wkh: false,
  },
  addresses: {
    np2wkh: null,
    p2wkh: null,
  },
}

// LND expects types to be sent as int, so this object will allow mapping from string to int
const ADDRESS_TYPES = {
  p2wkh: 0,
  np2wkh: 1,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * openWalletModal - Open wallet modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const openWalletModal = () => dispatch => dispatch(openModal('RECEIVE_MODAL'))

/**
 * closeWalletModal - Close wallet modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const closeWalletModal = () => dispatch => dispatch(closeModal('RECEIVE_MODAL'))

/**
 * newAddressSuccess - Generate new address success callback.
 *
 * @param {string} addressType Address type
 * @param {string} address Address
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const newAddressSuccess = (addressType, address) => async (dispatch, getState) => {
  const state = getState()
  const pubKey = state.info.data.identityPubkey

  // If we know the node's public key, store the address for reuse.
  if (pubKey) {
    const node = await window.db.nodes.get(pubKey)
    if (node) {
      node.setCurrentAddress(addressType, address)
      await window.db.nodes.put(node)
    } else {
      await window.db.nodes.put({ id: pubKey, addresses: { [addressType]: address } })
    }
  }

  dispatch({ type: NEW_ADDRESS_SUCCESS, addressType, address })
}

/**
 * createNewAddress - Creates a new address.
 *
 * @param {('p2wkh'|'np2wkh'|null)} addressType Address type. If not specified uses current saved config setting.
 * @returns {(dispatch:Function, getState:Function) => Promise<string>} Thunk
 */
export const createNewAddress = addressType => async (dispatch, getState) => {
  const getConfigAddressType = () => {
    const settings = settingsSelectors.currentConfig(getState())
    return settings && settings.address
  }
  const type = addressType || getConfigAddressType()
  const data = await grpc.services.Lightning.newAddress({ type: ADDRESS_TYPES[type] })
  return data.address
}

/**
 * newAddressFailure - Generate new address failure callback.
 *
 * @param {string} addressType Address type
 * @param {string} error Error message
 * @returns {(dispatch:Function) => void} Thunk
 */
export const newAddressFailure = (addressType, error) => dispatch => {
  const intl = getIntl()
  dispatch(
    showError(intl.formatMessage(messages.address_new_address_error, { addressType, error }))
  )
  dispatch({
    type: NEW_ADDRESS_FAILURE,
    addressType,
    error,
  })
}

/**
 * newAddress - Generate a new address and sets it as wallet current.
 *
 * @param {('p2wkh'|'np2wkh')} addressType Address type
 * @returns {(dispatch:Function) => Promise<string>} Thunk
 */
export const newAddress = addressType => async dispatch => {
  dispatch({ type: NEW_ADDRESS, addressType })
  try {
    const address = await dispatch(createNewAddress(addressType))
    await dispatch(newAddressSuccess(addressType, address))
    return address
  } catch (error) {
    dispatch(newAddressFailure(addressType, error.message))
    return null
  }
}

/**
 * initAddresses - Initialise addresses.
 *
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const initAddresses = () => async (dispatch, getState) => {
  try {
    dispatch({ type: FETCH_ADDRESSES })

    const state = getState()

    // Get node information (addresses are keyed under the node pubkey).
    const pubKey = state.info.data.identityPubkey
    const node = await window.db.nodes.get({ id: pubKey })

    // Get existing addresses for the node.
    const addresses = get(node, 'addresses', {})

    // Ensure that we have an address for all supported address types.
    const fetchAddresses = async types => {
      const data = {}
      await Promise.all(
        types.map(async type => {
          data[type] = addresses[type] || (await dispatch(newAddress(type)))
        })
      )
      return data
    }

    const allAddresses = await fetchAddresses(Object.keys(ADDRESS_TYPES))

    dispatch({ type: FETCH_ADDRESSES_SUCCESS, addresses: allAddresses })
  } catch (error) {
    dispatch({ type: FETCH_ADDRESSES_FAILURE, error: error.message })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_ADDRESSES]: state => {
    state.isAddressLoading = true
  },
  [FETCH_ADDRESSES_SUCCESS]: (state, { addresses }) => {
    state.addresses = addresses
    state.isAddressLoading = false
    state.fetchAddressError = null
  },
  [FETCH_ADDRESSES_FAILURE]: (state, { error }) => {
    state.isAddressLoading = false
    state.fetchAddressError = error
  },
  [NEW_ADDRESS]: (state, { addressType }) => {
    state.addressesLoading[addressType] = true
  },
  [NEW_ADDRESS_SUCCESS]: (state, { addressType, address }) => {
    state.addressesLoading[addressType] = false
    state.addresses[addressType] = address
  },
  [NEW_ADDRESS_FAILURE]: (state, { addressType }) => {
    state.addressesLoading[addressType] = false
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
