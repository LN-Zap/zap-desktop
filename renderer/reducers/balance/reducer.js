import Coin from '@zap/utils/coin'
import createReducer from '@zap/utils/createReducer'
import { grpc } from 'workers'

import * as constants from './constants'

const { FETCH_BALANCE, FETCH_BALANCE_SUCCESS, FETCH_BALANCE_FAILURE } = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isBalanceLoading Boolean indicating if balances are loading.
 * @property {string|null} walletBalance Total wallet balance.
 * @property {string|null} walletBalanceConfirmed Confirmed wallet balance.
 * @property {string|null} walletBalanceUnconfirmed Unconfirmed wallet balance.
 * @property {string|null} channelBalance Total channel balance.
 * @property {string|null} channelBalanceConfirmed Confirmed channel balance.
 * @property {string|null} channelBalancePending Pending channel balance.
 * @property {Error|null} fetchBalanceError Balance fetching error.
 */

/** @type {State} */
const initialState = {
  isBalanceLoading: false,
  walletBalance: null,
  walletBalanceConfirmed: null,
  walletBalanceUnconfirmed: null,
  channelBalance: null,
  channelBalanceConfirmed: null,
  channelBalancePending: null,
  fetchBalanceError: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * fetchBalance - Fetch balances.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const fetchBalance = () => async dispatch => {
  try {
    dispatch({ type: FETCH_BALANCE })
    const { walletBalance, channelBalance } = await grpc.services.Lightning.getBalance()
    dispatch({ type: FETCH_BALANCE_SUCCESS, walletBalance, channelBalance })
  } catch (error) {
    dispatch({ type: FETCH_BALANCE_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_BALANCE]: state => {
    state.isBalanceLoading = true
  },
  [FETCH_BALANCE_SUCCESS]: (state, { walletBalance, channelBalance }) => {
    state.isBalanceLoading = false
    state.walletBalance = walletBalance.totalBalance
    state.walletBalanceConfirmed = walletBalance.confirmedBalance
    state.walletBalanceUnconfirmed = walletBalance.unconfirmedBalance
    state.channelBalance = Coin(channelBalance.balance)
      .add(Coin(channelBalance.pendingOpenBalance))
      .toString()
    state.channelBalanceConfirmed = channelBalance.balance
    state.channelBalancePending = channelBalance.pendingOpenBalance
  },
  [FETCH_BALANCE_FAILURE]: (state, { error }) => {
    state.isBalanceLoading = false
    state.fetchBalanceError = error
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
