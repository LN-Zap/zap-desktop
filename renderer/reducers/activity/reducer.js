import { send } from 'redux-electron-ipc'

import { getIntl } from '@zap/i18n'
import createReducer from '@zap/utils/createReducer'
import { mainLog } from '@zap/utils/log'
import { fetchBalance } from 'reducers/balance'
import { fetchChannels } from 'reducers/channels'
import { fetchInfo, infoSelectors } from 'reducers/info'
import { receiveInvoices } from 'reducers/invoice'
import { openModal, closeModal } from 'reducers/modal'
import { showError, showNotification } from 'reducers/notification'
import { receivePayments } from 'reducers/payment'
import { settingsSelectors } from 'reducers/settings'
import { receiveTransactions } from 'reducers/transaction'
import { walletSelectors } from 'reducers/wallet'

import * as constants from './constants'
import messages from './messages'
import { hasNextPage, isPageLoading } from './selectors'
import { createActivityPaginator } from './utils'

const {
  SET_ACTIVITY_MODAL,
  HIDE_ACTIVITY_MODAL,
  CHANGE_FILTER,
  UPDATE_SEARCH_TEXT,
  FETCH_ACTIVITY_HISTORY,
  FETCH_ACTIVITY_HISTORY_SUCCESS,
  FETCH_ACTIVITY_HISTORY_FAILURE,
  ADD_FILTER,
  SET_HAS_NEXT_PAGE,
  SET_ERROR_DIALOG_DETAILS,
  SET_PAGE_LOADING,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {Set<string>} filter Currently applied activity filter.
 * @property {object[]} filters List of supported activity filters.
 * @property {{itemType: string|null, itemId: string|null}} modal Activity item currently showing in activity modal.
 * @property {Error|null} errorDialogDetails Activity error currently showing in error details dialog.
 * @property {string|null} searchText Current activity search string.
 * @property {boolean} isActivityLoading Boolean indicating if activity is loading.
 * @property {Error|null} activityLoadingError Error from loading activity.
 * @property {boolean} hasNextPage Boolean indicating if there are more activity pages to fetch.
 * @property {boolean} isPageLoading Boolean indicating if the activity page is loading.
 */

/** @type {State} */
const initialState = {
  filter: new Set([...constants.defaultFilter]),
  filters: [
    { key: 'SENT_ACTIVITY' },
    { key: 'RECEIVED_ACTIVITY' },
    { key: 'PENDING_ACTIVITY' },
    { key: 'EXPIRED_ACTIVITY' },
    { key: 'INTERNAL_ACTIVITY' },
  ],
  modal: {
    itemType: null,
    itemId: null,
  },
  errorDialogDetails: null,
  searchText: null,
  isActivityLoading: false,
  activityLoadingError: null,
  hasNextPage: true,
  isPageLoading: false,
}

// ------------------------------------
// Helpers
// ------------------------------------

// activity paginator object. must be reset for each wallet login
/** @type {Function | null} */
let paginator = null

/**
 * getPaginator - Returns current activity paginator object. This acts as a singleton
 * and creates paginator if it's not initialized.
 *
 * @param {{
 *  invoiceHandler: function,
 *  paymentHandler: function,
 *  transactionHandler: function
 * }} handlers Pagination handlers
 * @returns {Function} Paginator
 */
export const getPaginator = handlers => {
  if (!paginator) {
    paginator = createActivityPaginator(handlers)
  }
  return paginator
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * setPageLoading - Set boolean indicating if page is in the process of loading.
 *
 * @param {boolean} isPageLoading Boolean indicating if page is loading
 * @returns {object} Action
 */
export const setPageLoading = isPageLoading => {
  return {
    type: SET_PAGE_LOADING,
    isPageLoading,
  }
}

/**
 * setErorDialogDetails - Set the error dialog details.
 *
 * @param {string} details Error details
 * @returns {object} Action
 */
export const setErorDialogDetails = details => {
  return {
    type: SET_ERROR_DIALOG_DETAILS,
    details,
  }
}

/**
 * changeFilter - Toggle specified activity filter.
 *
 * @param {...string} filterList Filter to apply
 * @returns {object} Action
 */
export function changeFilter(...filterList) {
  return {
    type: CHANGE_FILTER,
    filterList,
  }
}
/**
 * addFilter - Enable specified activity filters.
 *
 * @param {...string} filterList Filters to apply
 * @returns {object} Action
 */
export function addFilter(...filterList) {
  return {
    type: ADD_FILTER,
    filterList,
  }
}

/**
 * updateSearchText - Set the current activity search string.
 *
 * @param {string|null} searchText Search string to apply
 * @returns {object} Action
 */
export function updateSearchText(searchText = null) {
  return {
    type: UPDATE_SEARCH_TEXT,
    searchText,
  }
}

/**
 * saveInvoice - Initiates saving of invoice pdf.
 *
 * @param {object} options invoice options
 * @param {string} options.defaultFilename invoice title
 * @param {string} options.title invoice title
 * @param {string} options.subtitle invoice subtitle
 * @param {Array<object>} options.invoiceData invoice rows
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const saveInvoice = ({
  defaultFilename,
  title,
  subtitle,
  invoiceData,
}) => async dispatch => {
  dispatch(send('saveInvoice', { defaultFilename, title, subtitle, invoiceData }))
}

/**
 * saveInvoiceFailure - Invoice save failed event.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const saveInvoiceFailure = () => dispatch => {
  dispatch(showError(getIntl().formatMessage(messages.activity_invoice_download_error)))
}

/**
 * saveInvoiceSuccess - Invoice save success event.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const saveInvoiceSuccess = () => dispatch => {
  dispatch(showNotification(getIntl().formatMessage(messages.activity_invoice_download_success)))
}

/**
 * setActivityModal - Show the activity modal with a given activity item.
 *
 * @param {string} itemType Item type
 * @param {string} itemId Item id
 * @returns {(dispatch:Function) => void} Thunk
 */
export const setActivityModal = (itemType, itemId) => dispatch => {
  dispatch({ type: SET_ACTIVITY_MODAL, itemType, itemId })
}

/**
 * showActivityModal - Show the activity modal with a given activity item.
 *
 * @param {string} itemType Item type
 * @param {string} itemId Item id
 * @returns {(dispatch:Function) => void} Thunk
 */
export const showActivityModal = (itemType, itemId) => dispatch => {
  dispatch(setActivityModal(itemType, itemId))
  dispatch(openModal('ACTIVITY_MODAL'))
}

/**
 * hideActivityModal - Hide the activity modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const hideActivityModal = () => dispatch => {
  dispatch({ type: HIDE_ACTIVITY_MODAL })
  dispatch(closeModal('ACTIVITY_MODAL'))
}

/**
 * resetActivity - Reset activity history.
 *
 * @returns {() => void} Thunk
 */
export const resetActivity = () => () => {
  paginator = null
}

/**
 * loadPage - Loads next activity page if it's available.
 *
 * @param {boolean} reload Boolean indicating wether to load first page.
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const loadPage = (reload = false) => async (dispatch, getState) => {
  if (isPageLoading(getState())) {
    return
  }
  dispatch(setPageLoading(true))

  await dispatch(fetchInfo())
  const state = getState()
  const config = settingsSelectors.currentConfig(state)
  const blockHeight = infoSelectors.blockHeight(state)
  const activeWallet = walletSelectors.activeWallet(state)

  // Checks to see if the currently active wallet is the same one as when the page load started.
  const isStllActiveWallet = () => walletSelectors.activeWallet(getState()) === activeWallet

  // It's possible that the grpc response is received after the wallet has been disconnected.
  // See https://github.com/grpc/grpc-node/issues/1603
  // Ensure that the items received belong to the currently active wallet.
  const shouldUpdate = items => items && isStllActiveWallet()

  const handlers = {
    invoiceHandler: items => shouldUpdate(items) && dispatch(receiveInvoices(items)),
    paymentHandler: items => shouldUpdate(items) && dispatch(receivePayments(items)),
    transactionHandler: items => shouldUpdate(items) && dispatch(receiveTransactions(items)),
  }
  const thisPaginator = reload ? createActivityPaginator(handlers) : getPaginator(handlers)

  if (hasNextPage(getState())) {
    const { pageSize } = config.activity
    const { hasNextPage } = await thisPaginator(pageSize, blockHeight)
    if (!isStllActiveWallet()) {
      return
    }
    dispatch({ type: SET_HAS_NEXT_PAGE, value: hasNextPage })
  }

  dispatch(setPageLoading(false))
}

/**
 * reloadActivityHead - Reloads activity head.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const reloadActivityHead = () => async dispatch => {
  mainLog.debug(`reloading activity pages`)
  await dispatch(loadPage(true))
}

/**
 * initActivityHistory - Load user activity history.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const initActivityHistory = () => async dispatch => {
  dispatch({ type: FETCH_ACTIVITY_HISTORY })
  try {
    await Promise.all([dispatch(loadPage()), dispatch(fetchChannels()), dispatch(fetchBalance())])
    dispatch({ type: FETCH_ACTIVITY_HISTORY_SUCCESS })
  } catch (error) {
    dispatch({ type: FETCH_ACTIVITY_HISTORY_FAILURE, error })
  }
}

/**
 * reloadActivityHistory - Reload activity history.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const reloadActivityHistory = () => async dispatch => {
  dispatch({ type: FETCH_ACTIVITY_HISTORY })
  try {
    await Promise.all([
      dispatch(reloadActivityHead()),
      dispatch(fetchChannels()),
      dispatch(fetchBalance()),
    ])
    dispatch({ type: FETCH_ACTIVITY_HISTORY_SUCCESS })
  } catch (error) {
    dispatch({ type: FETCH_ACTIVITY_HISTORY_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [SET_ACTIVITY_MODAL]: (state, { itemType, itemId }) => {
    state.modal = { itemType, itemId }
  },
  [HIDE_ACTIVITY_MODAL]: state => {
    state.modal = { itemType: null, itemId: null }
  },
  [CHANGE_FILTER]: (state, { filterList }) => {
    const { filter } = state
    filterList.forEach(item => {
      if (filter.has(item)) {
        filter.delete(item)
      } else {
        filter.add(item)
      }
    })
  },
  [ADD_FILTER]: (state, { filterList }) => {
    const { filter } = state
    // No filters means all filters are active so no need to add anything
    if (filter.size > 0) {
      filterList.forEach(item => filter.add(item))
    }
  },
  [UPDATE_SEARCH_TEXT]: (state, { searchText }) => {
    state.searchText = searchText
  },
  [FETCH_ACTIVITY_HISTORY]: state => {
    state.isActivityLoading = true
  },
  [FETCH_ACTIVITY_HISTORY_SUCCESS]: state => {
    state.isActivityLoading = false
  },
  [FETCH_ACTIVITY_HISTORY_FAILURE]: (state, { error }) => {
    state.isActivityLoading = false
    state.activityLoadingError = error
  },
  [SET_HAS_NEXT_PAGE]: (state, { value }) => {
    state.hasNextPage = value
  },
  [SET_ERROR_DIALOG_DETAILS]: (state, { details }) => {
    state.errorDialogDetails = details
  },
  [SET_PAGE_LOADING]: (state, { isPageLoading }) => {
    state.isPageLoading = isPageLoading
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
