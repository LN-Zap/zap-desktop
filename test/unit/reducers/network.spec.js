import reducer, {
  GET_DESCRIBE_NETWORK,
  RECEIVE_DESCRIBE_NETWORK,
  UPDATE_NODE_DATA,
} from 'reducers/network'

import snapshotDiff from '../__helpers__/snapshotDiff'

describe('reducers', () => {
  describe('networkReducer', () => {
    it('should handle initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle GET_DESCRIBE_NETWORK', () => {
      const action = {
        type: GET_DESCRIBE_NETWORK,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RECEIVE_DESCRIBE_NETWORK', () => {
      const action = {
        type: RECEIVE_DESCRIBE_NETWORK,
        nodes: [{ some: 'data' }],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle UPDATE_NODE_DATA', () => {
      const action = {
        type: UPDATE_NODE_DATA,
        data: [{ some: 'data' }],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })
  })
})
