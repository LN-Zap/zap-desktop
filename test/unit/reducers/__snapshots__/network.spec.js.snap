// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers networkReducer should handle GET_DESCRIBE_NETWORK 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
-   "networkLoading": false,
+   "networkLoading": true,
    "nodes": Array [],
  }
`;

exports[`reducers networkReducer should handle RECEIVE_DESCRIBE_NETWORK 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "networkLoading": false,
-   "nodes": Array [],
+   "nodes": Array [
+     Object {
+       "some": "data",
+     },
+   ],
  }
`;

exports[`reducers networkReducer should handle UPDATE_NODE_DATA 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "networkLoading": false,
-   "nodes": Array [],
+   "nodes": Array [
+     Object {
+       "some": "data",
+     },
+   ],
  }
`;

exports[`reducers networkReducer should handle initial state 1`] = `
Object {
  "networkLoading": false,
  "nodes": Array [],
}
`;
